easybuild/easyblocks/f/freefem.py

```diff
@@ -0,0 +1,109 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for FreeFem++, implemented as an easyblock
+
+@author: Balazs Hajgato (Free University Brussels - VUB)
+"""
+import os
+
+from easybuild.easyblocks.generic.configuremake import ConfigureMake
+from easybuild.tools.run import run_cmd
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.modules import get_software_root
+
+
+class EB_FreeFem_plus__plus_(ConfigureMake):
+    """Support for building and installing FreeFem++."""
+
+    def configure_step(self):
+        """FreeFem++ configure should run twice. First PETSc configured, then PETSc have to be build,
+        then configure FreeFem++ with the builded PETSc."""
+
+        # first Autoreconf has to be run
+        if not get_software_root('Autoconf'):
+            raise EasyBuildError("Autoconf is required to build FreeFem++. Please add it as build dependency")
+
+        run_cmd("autoreconf -i", log_all=True, simple=False)
+
+        # delete old installation, then set keeppreviousinstall to True (do not delete PETsc install)
+        self.make_installdir()
+        self.cfg['keeppreviousinstall'] = True
+
+        # configure and make petsc-slepc
+        cmd = "./configure --prefix=%s &&" % self.installdir
+        cmd += "cd download/ff-petsc &&"
+        cmd += "make petsc-slepc &&"
+        cmd += "cd ../.."
+        run_cmd(cmd, log_all=True, simple=False)
+
+        # check HDF5 and GSL deps and add the corresponding configopts
+        hdf5_root = get_software_root('HDF5')
+        if hdf5_root:
+            self.cfg.update('configopts', '--with-hdf5=%s ' % os.path.join(hdf5_root, 'bin', 'h5pcc'))
+
+        gsl_root = get_software_root('GSL')
+        if gsl_root:
+            self.cfg.update('configopts', '--with-gsl-include=%s ' % os.path.join(gsl_root, 'include'))
+            self.cfg.update('configopts', '--with-gsl-ldflags=%s ' % os.path.join(gsl_root, 'lib'))
+
+        # We should download deps
+        self.cfg.update('configopts', '--enable-download ')
+
+        # PaStiX does not work parallel, so disable it.
+        self.cfg.update('configopts', '--disable-pastix ')
+
+        super(EB_FreeFem_plus__plus_, self).configure_step()
+
+    def build_step(self):
+        # dependencies are heavily patched by FreeFem++, we therefore let it download and install them itself
+        self.cfg.update('prebuildopts', 'download/getall -a -o ScaLAPACK,ARPACK,freeYams,Gmm++,Hips,Ipopt,METIS,'
+                        'ParMETIS,MMG3D,mshmet,MUMPS,NLopt,pARMS,PaStiX,Scotch,SuiteSparse,SuperLU_DIST,SuperLU,'
+                        'TetGen,PETSc,SLEPc,hpddm &&')
```

Comment:
Hmmm, the first thing you do in configure_step is build ff_petsc, but here you ask it to download petsc and slepc again.

Reply:
Indeed, but anyway, it checks whether things are already downloaded or not, so `PETSc` and `SLEPc` is not downloaded again. I do not remember why I did build the `PETSc` and `SLEPc` in advance, but maybe some of the downloaded modules needs it at compilation time, and this is the way how I made sure that they are already built.

If you want, I can remove `PETSc` and `SLEPc`, but anyway, they would not be downloaded again.